### ISSUE TITLE
Move to line-based diff of sources and outputs

### ISSUE THREAD

**Issue by martinal** [jupyter/nbdime#17]
The current implementation does a charbased diff of sources, we don't want that. The same applies to some other strings, in particular some of the output fields. However there are many single line/word strings we want to deal with as atomic values. How do we distinguish which is which?

The easiest way to implement this is to
1) split lines for all relevant strings in notebooks when loaded
2) treat all strings as atomic values in the diff algorithm
3) rejoin lines when done with notebooks

@minrk Can I use split_lines/rejoin_lines from nbformat.v4.rwbase? It looks like they do what I want.

**Comment by minrk** [comment]
I wouldn't use that, it's a private utility function, and it's only one line. You can use stdlib str methods to do it anyway:

```python
line_list = multiline.splitlines(True) # True preserves newline characters
...
multiline = u''.join(line_list) # using True above means join with '' instead of '\n'
```

Unlike nbformat, which *allows* strings or multi-line lists as input, when you use nbformat to read, you are guaranteed to only receive strings, so there's no ambiguity to handle.

**Comment by martinal** [comment]
It's not the line splitting I'm after, it's which fields to split.

Also, I'm not sure where to do this splitting, because if I do it internally in diff then applying patch to the original notebook will not work, so the ambiguity is very much present.

**Comment by minrk** [comment]
oh, gotcha. I'm not sure, then. I might still have our own implementation here, possibly inside diff computations.

**Comment by martinal** [comment]
One approach would be to leave the splitting outside the notebook diff and merge algorithms. If you want diff with the cell sources treated as large atomic values, pass a notebook without sources split. You get back a diff consistent with what you pass in. Do that with merge and there will be no attempt at merging lines within cells, and conflicts will occur at cell level.

**Comment by minrk** [comment]
I think that's a fine place to start.

**Comment by vidartf** [comment]
Implemented in #76. Currently the server only splits strings for the purpose of diffing them, but all string diffs treat the string as a sequence of lines. There is also a utility function `flatten_list_of_string_diff` to convert such a diff to diffs that work directly on the multi-line string.
